dll/win32/shell32/folders/CDrivesFolder.cpp

```diff
@@ -57,6 +58,85 @@ static int iDriveTypeIds[7] = { IDS_DRIVE_FIXED,       /* DRIVE_UNKNOWN */
 *   IShellFolder implementation
 */
 
+static BOOL TryToLockDrive(HANDLE hDrive, DWORD dwRetryCount, DWORD dwSleep)
+{
+    DWORD dwError, dwBytesReturned;
+    for (DWORD i = 0; i < dwRetryCount; ++i)
+    {
+        if (DeviceIoControl(hDrive, FSCTL_LOCK_VOLUME, NULL, 0, NULL, 0, &dwBytesReturned, NULL))
+            return TRUE;
+
+        dwError = GetLastError();
+        if (dwError == ERROR_INVALID_FUNCTION)
+            break; /* don't sleep if function is not implemented */
+
+        Sleep(dwSleep);
+    }
+    SetLastError(dwError);
+    return FALSE;
+}
+
+// NOTE: See also https://support.microsoft.com/en-us/help/165721/how-to-ejecting-removable-media-in-windows-nt-windows-2000-windows-xp
+static BOOL DoEjectDrive(const WCHAR *physical, UINT nDriveType, INT& nStringID)
+{
+    DWORD dwAccessMode, dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE;
+    if (nDriveType == DRIVE_REMOVABLE)
+        dwAccessMode = GENERIC_READ | GENERIC_WRITE;
+    else if (nDriveType == DRIVE_CDROM)
+        dwAccessMode = GENERIC_READ;
+
+    HANDLE hDrive = CreateFile(physical, dwAccessMode, dwShareMode, 0, OPEN_EXISTING, 0, NULL);
+    if (hDrive == INVALID_HANDLE_VALUE)
+        return FALSE;
+
+    BOOL bResult;
+    DWORD dwBytesReturned, dwError = NO_ERROR;
+    do
+    {
+        PREVENT_MEDIA_REMOVAL removal;
+        bResult = TryToLockDrive(hDrive, 3, 250);
+        if (!bResult)
+        {
+            dwError = GetLastError();
+            if (dwError != ERROR_INVALID_FUNCTION)  /* temporary hack */
+            {
+                nStringID = IDS_CANTLOCKVOLUME; /* Unable to lock volume */
+                break;
+            }
+        }
+        bResult = DeviceIoControl(hDrive, FSCTL_DISMOUNT_VOLUME, NULL, 0, NULL, 0, &dwBytesReturned, NULL);
+        if (!bResult)
+        {
+            dwError = GetLastError();
+            if (dwError != ERROR_INVALID_FUNCTION)  /* temporary hack */
+            {
+                nStringID = IDS_CANTDISMOUNTVOLUME; /* Unable to dismount volume */
+                break;
```

Comment:
You cannot just break here. You have to unlock volume! You're leaking a volume lock (dramatic!)

Reply:
Ok.